Panic due to the nil PD/TSO client

## Bug Report

Ref https://github.com/tikv/client-go/actions/runs/4403635107/jobs/7712245145.

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0xcce915]

goroutine 5994 [running]:
github.com/tikv/pd/client.(*client).GetRegion(0xc000cda000, {0x20df860, 0xc0020a3740}, {0xc000c9a0a0, 0x9, 0x9}, {0xc000c9e2e8, 0x1, 0x493301?})
	/home/runner/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20230313083840-3e3ae55f68ac/client.go:596 +0x3b5
github.com/tikv/client-go/v2/internal/locate.(*CodecPDClient).GetRegion(0xc000645860, {0x20df860, 0xc0020a3740}, {0xc0020b013a?, 0x1e61540?, 0x40f7ff?}, {0xc000c9e2e8, 0x1, 0x1})
	/home/runner/work/client-go/client-go/internal/locate/pd_codec.go:97 +0xa8
github.com/tikv/client-go/v2/internal/locate.(*RegionCache).loadRegion(0xc000fd8e00, 0xc0007ed050, {0xc0020b013a, 0x2, 0x2}, 0x0)
	/home/runner/work/client-go/client-go/internal/locate/region_cache.go:1502 +0x388
github.com/tikv/client-go/v2/internal/locate.(*RegionCache).findRegionByKey(0xc000fd8e00, 0xc0007ed050, {0xc0020b013a, 0x2, 0x2}, 0xc0?)
	/home/runner/work/client-go/client-go/internal/locate/region_cache.go:986 +0x425
github.com/tikv/client-go/v2/internal/locate.(*RegionCache).LocateKey(0x19?, 0x0?, {0xc0020b013a?, 0x0?, 0x1b025c0?})
	/home/runner/work/client-go/client-go/internal/locate/region_cache.go:941 +0x36
github.com/tikv/client-go/v2/txnkv/transaction.groupSortedMutationsByRegion(0x0?, 0x0?, {0x20f2740, 0xc0020c1080})
	/home/runner/work/client-go/client-go/txnkv/transaction/2pc.go:825 +0x2d0
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).groupMutations(0xc00056d340, 0xc0007ed050, {0x20f2740, 0xc0020c1080})
	/home/runner/work/client-go/client-go/txnkv/transaction/2pc.go:842 +0x76
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).doActionOnMutations(0xc00056d340, 0xc000cf5520?, {0x20de590?, 0x365f0a0}, {0x20f2740, 0xc0020c1080})
	/home/runner/work/client-go/client-go/txnkv/transaction/2pc.go:791 +0x6c
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).pessimisticRollbackMutations(0xc0007ed050?, 0xc0000ba6c0?, {0x20f2740?, 0xc0020c1080?})
	/home/runner/work/client-go/client-go/txnkv/transaction/pessimistic.go:563 +0x17e
github.com/tikv/client-go/v2/txnkv/transaction.actionPessimisticRollback.handleSingleBatch({0x0}, 0xc00056d340, 0xc0007ed050, {{0x4, 0x1, 0x5}, {0x20f2740, 0xc0020c1080}, 0x1})
	/home/runner/work/client-go/client-go/txnkv/transaction/pessimistic.go:527 +0x2f1
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).doActionOnBatches(0xc00056d340, 0xc0007ed050, {0x20de590?, 0x365f0a0}, {0xc0020bec30?, 0x1, 0xc000090a38?})
	/home/runner/work/client-go/client-go/txnkv/transaction/2pc.go:1056 +0x32f
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).doActionOnGroupMutations(0xc00056d340, 0xc0007ed050, {0x20de590, 0x365f0a0?}, {0xc0020bec00, 0x1, 0xc000090d80?})
	/home/runner/work/client-go/client-go/txnkv/transaction/2pc.go:1037 +0x1317
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).doActionOnMutations(0xc00056d340, 0x1a7ad20?, {0x20de590?, 0x365f0a0}, {0x20f2740, 0xc0020c0fc0})
	/home/runner/work/client-go/client-go/txnkv/transaction/2pc.go:800 +0xc7
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).pessimisticRollbackMutations(0xc000c95440?, 0x1dad5eb?, {0x20f2740?, 0xc0020c0fc0?})
	/home/runner/work/client-go/client-go/txnkv/transaction/pessimistic.go:563 +0x17e
github.com/tikv/client-go/v2/txnkv/transaction.(*KVTxn).asyncPessimisticRollback.func1()
	/home/runner/work/client-go/client-go/txnkv/transaction/txn.go:1307 +0x5e5
created by github.com/tikv/client-go/v2/txnkv/transaction.(*KVTxn).asyncPessimisticRollback
	/home/runner/work/client-go/client-go/txnkv/transaction/txn.go:1290 +0x38a
```

The client here may return a nil pointer which will cause panic, we should prevent this case.

https://github.com/tikv/pd/blob/0b362383369e774aa238f119c0f8b6c371c56339/client/client.go#L478-L483

--- a/client/client.go
+++ b/client/client.go
@@ -466,7 +466,12 @@ func (c *client) GetAllMembers(ctx context.Context) ([]*pdpb.Member, error) {
 	ctx, cancel := context.WithTimeout(ctx, c.option.timeout)
 	req := &pdpb.GetMembersRequest{Header: c.requestHeader()}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().GetMembers(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.GetMembers(ctx, req)
 	cancel()
 	if err = c.respForErr(cmdFailDurationGetAllMembers, start, err, resp.GetHeader()); err != nil {
 		return nil, err
@@ -593,7 +598,12 @@ func (c *client) GetRegion(ctx context.Context, key []byte, opts ...GetRegionOpt
 		NeedBuckets: options.needBuckets,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().GetRegion(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.GetRegion(ctx, req)
 	cancel()
 
 	if err = c.respForErr(cmdFailDurationGetRegion, start, err, resp.GetHeader()); err != nil {
@@ -663,7 +673,12 @@ func (c *client) GetPrevRegion(ctx context.Context, key []byte, opts ...GetRegio
 		NeedBuckets: options.needBuckets,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().GetPrevRegion(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.GetPrevRegion(ctx, req)
 	cancel()
 
 	if err = c.respForErr(cmdFailDurationGetPrevRegion, start, err, resp.GetHeader()); err != nil {
@@ -691,7 +706,12 @@ func (c *client) GetRegionByID(ctx context.Context, regionID uint64, opts ...Get
 		NeedBuckets: options.needBuckets,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().GetRegionByID(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.GetRegionByID(ctx, req)
 	cancel()
 
 	if err = c.respForErr(cmdFailedDurationGetRegionByID, start, err, resp.GetHeader()); err != nil {
@@ -721,7 +741,12 @@ func (c *client) ScanRegions(ctx context.Context, key, endKey []byte, limit int)
 		Limit:    int32(limit),
 	}
 	scanCtx = grpcutil.BuildForwardContext(scanCtx, c.GetLeaderAddr())
-	resp, err := c.getClient().ScanRegions(scanCtx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.ScanRegions(scanCtx, req)
 
 	if err = c.respForErr(cmdFailedDurationScanRegions, start, err, resp.GetHeader()); err != nil {
 		return nil, err
@@ -772,7 +797,12 @@ func (c *client) GetStore(ctx context.Context, storeID uint64) (*metapb.Store, e
 		StoreId: storeID,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().GetStore(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.GetStore(ctx, req)
 	cancel()
 
 	if err = c.respForErr(cmdFailedDurationGetStore, start, err, resp.GetHeader()); err != nil {
@@ -812,7 +842,12 @@ func (c *client) GetAllStores(ctx context.Context, opts ...GetStoreOption) ([]*m
 		ExcludeTombstoneStores: options.excludeTombstone,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().GetAllStores(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.GetAllStores(ctx, req)
 	cancel()
 
 	if err = c.respForErr(cmdFailedDurationGetAllStores, start, err, resp.GetHeader()); err != nil {
@@ -835,7 +870,12 @@ func (c *client) UpdateGCSafePoint(ctx context.Context, safePoint uint64) (uint6
 		SafePoint: safePoint,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().UpdateGCSafePoint(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return 0, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.UpdateGCSafePoint(ctx, req)
 	cancel()
 
 	if err = c.respForErr(cmdFailedDurationUpdateGCSafePoint, start, err, resp.GetHeader()); err != nil {
@@ -865,7 +905,12 @@ func (c *client) UpdateServiceGCSafePoint(ctx context.Context, serviceID string,
 		SafePoint: safePoint,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().UpdateServiceGCSafePoint(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return 0, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.UpdateServiceGCSafePoint(ctx, req)
 	cancel()
 
 	if err = c.respForErr(cmdFailedDurationUpdateServiceGCSafePoint, start, err, resp.GetHeader()); err != nil {
@@ -893,7 +938,12 @@ func (c *client) scatterRegionsWithGroup(ctx context.Context, regionID uint64, g
 		Group:    group,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().ScatterRegion(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.ScatterRegion(ctx, req)
 	cancel()
 	if err != nil {
 		return err
@@ -933,7 +983,12 @@ func (c *client) SplitAndScatterRegions(ctx context.Context, splitKeys [][]byte,
 	}
 
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	return c.getClient().SplitAndScatterRegions(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	return protoClient.SplitAndScatterRegions(ctx, req)
 }
 
 func (c *client) GetOperator(ctx context.Context, regionID uint64) (*pdpb.GetOperatorResponse, error) {
@@ -951,7 +1006,12 @@ func (c *client) GetOperator(ctx context.Context, regionID uint64) (*pdpb.GetOpe
 		RegionId: regionID,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	return c.getClient().GetOperator(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	return protoClient.GetOperator(ctx, req)
 }
 
 // SplitRegions split regions by given split keys
@@ -974,7 +1034,12 @@ func (c *client) SplitRegions(ctx context.Context, splitKeys [][]byte, opts ...R
 		RetryLimit: options.retryLimit,
 	}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	return c.getClient().SplitRegions(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	return protoClient.SplitRegions(ctx, req)
 }
 
 func (c *client) requestHeader() *pdpb.RequestHeader {
@@ -999,7 +1064,12 @@ func (c *client) scatterRegionsWithOptions(ctx context.Context, regionsID []uint
 	}
 
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().ScatterRegion(ctx, req)
+	protoClient := c.getClient()
+	if protoClient == nil {
+		cancel()
+		return nil, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.ScatterRegion(ctx, req)
 	cancel()
 
 	if err != nil {
@@ -1037,7 +1107,11 @@ func trimHTTPPrefix(str string) string {
 }
 
 func (c *client) LoadGlobalConfig(ctx context.Context, names []string, configPath string) ([]GlobalConfigItem, int64, error) {
-	resp, err := c.getClient().LoadGlobalConfig(ctx, &pdpb.LoadGlobalConfigRequest{Names: names, ConfigPath: configPath})
+	protoClient := c.getClient()
+	if protoClient == nil {
+		return nil, 0, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.LoadGlobalConfig(ctx, &pdpb.LoadGlobalConfigRequest{Names: names, ConfigPath: configPath})
 	if err != nil {
 		return nil, 0, err
 	}
@@ -1062,7 +1136,11 @@ func (c *client) StoreGlobalConfig(ctx context.Context, configPath string, items
 	for i, it := range items {
 		resArr[i] = &pdpb.GlobalConfigItem{Name: it.Name, Value: it.Value, Kind: it.EventType, Payload: it.PayLoad}
 	}
-	_, err := c.getClient().StoreGlobalConfig(ctx, &pdpb.StoreGlobalConfigRequest{Changes: resArr, ConfigPath: configPath})
+	protoClient := c.getClient()
+	if protoClient == nil {
+		return errs.ErrClientGetProtoClient
+	}
+	_, err := protoClient.StoreGlobalConfig(ctx, &pdpb.StoreGlobalConfigRequest{Changes: resArr, ConfigPath: configPath})
 	if err != nil {
 		return err
 	}
@@ -1073,7 +1151,11 @@ func (c *client) WatchGlobalConfig(ctx context.Context, configPath string, revis
 	// TODO: Add retry mechanism
 	// register watch components there
 	globalConfigWatcherCh := make(chan []GlobalConfigItem, 16)
-	res, err := c.getClient().WatchGlobalConfig(ctx, &pdpb.WatchGlobalConfigRequest{
+	protoClient := c.getClient()
+	if protoClient == nil {
+		return nil, errs.ErrClientGetProtoClient
+	}
+	res, err := protoClient.WatchGlobalConfig(ctx, &pdpb.WatchGlobalConfigRequest{
 		ConfigPath: configPath,
 		Revision:   revision,
 	})
@@ -1115,7 +1197,11 @@ func (c *client) WatchGlobalConfig(ctx context.Context, configPath string, revis
 }
 
 func (c *client) GetExternalTimestamp(ctx context.Context) (uint64, error) {
-	resp, err := c.getClient().GetExternalTimestamp(ctx, &pdpb.GetExternalTimestampRequest{
+	protoClient := c.getClient()
+	if protoClient == nil {
+		return 0, errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.GetExternalTimestamp(ctx, &pdpb.GetExternalTimestampRequest{
 		Header: c.requestHeader(),
 	})
 	if err != nil {
@@ -1129,7 +1215,11 @@ func (c *client) GetExternalTimestamp(ctx context.Context) (uint64, error) {
 }
 
 func (c *client) SetExternalTimestamp(ctx context.Context, timestamp uint64) error {
-	resp, err := c.getClient().SetExternalTimestamp(ctx, &pdpb.SetExternalTimestampRequest{
+	protoClient := c.getClient()
+	if protoClient == nil {
+		return errs.ErrClientGetProtoClient
+	}
+	resp, err := protoClient.SetExternalTimestamp(ctx, &pdpb.SetExternalTimestampRequest{
 		Header:    c.requestHeader(),
 		Timestamp: timestamp,
 	})

--- a/client/errs/errno.go
+++ b/client/errs/errno.go
@@ -30,6 +30,7 @@ const (
 
 // client errors
 var (
+	ErrClientGetProtoClient   = errors.Normalize("failed to get proto client, %s", errors.RFCCodeText("PD:client:ErrClientGetProtoClient"))
 	ErrClientCreateTSOStream  = errors.Normalize("create TSO stream failed, %s", errors.RFCCodeText("PD:client:ErrClientCreateTSOStream"))
 	ErrClientGetTSOTimeout    = errors.Normalize("get TSO timeout", errors.RFCCodeText("PD:client:ErrClientGetTSOTimeout"))
 	ErrClientGetTSO           = errors.Normalize("get TSO failed, %v", errors.RFCCodeText("PD:client:ErrClientGetTSO"))

